### ISSUE TITLE
Quoine: Authentication failed

### ISSUE THREAD

**Issue by wuqianming** [timmolter/XChange#1398]
Caught: org.knowm.xchange.exceptions.ExchangeException: API Authentication failed
org.knowm.xchange.exceptions.ExchangeException: API Authentication failed
	at org.knowm.xchange.quoine.service.QuoineAccountServiceRaw.getQuoineFiatAccountInfo(QuoineAccountServiceRaw.java:32)
	at org.knowm.xchange.quoine.service.QuoineAccountService.getAccountInfo(QuoineAccountService.java:45)
	at org.knowm.xchange.service.account.AccountService$getAccountInfo.call(Unknown Source)

**Comment by jheusser** [comment]
@wuqianming currently you need to set KEY_TOKEN_ID and KEY_USER_SECRET in the exchange specific parameters. The two values you get from the API key you generate on Quoine.

```
exchange.getExchangeSpecification().getExchangeSpecificParameters().put(QuoineExchange.KEY_TOKEN_ID), "your token id");
exchange.getExchangeSpecification().getExchangeSpecificParameters().put(QuoineExchange.KEY_USER_SECRET, "your secret");
```
I will make this more user friendly.

**Comment by wuqianming** [comment]
@jheusser  I've tried but got this,seems like it's a readonly property:

Caught: groovy.lang.ReadOnlyPropertyException: Cannot set readonly property: KEY_TOKEN_ID for class: org.knowm.xchange.quoine.QuoineExchange
groovy.lang.ReadOnlyPropertyException: Cannot set readonly property: KEY_TOKEN_ID for class: org.knowm.xchange.quoine.QuoineExchange

**Comment by wuqianming** [comment]
Thank you so much man, I've figured it out.
BTW may I have your e-mail address ? @jheusser

**Comment by npomfret** [comment]
I'm getting a 401 too.  Is there a trick to this? I've tried:

```
specification.setExchangeSpecificParametersItem(QuoineExchange.KEY_TOKEN_ID, keys.apiKey);
specification.setExchangeSpecificParametersItem(QuoineExchange.KEY_USER_SECRET, keys.secret);
```

**Comment by jheusser** [comment]
@npomfret This is correct. How do you instantiate the `Exchange` instance?

**Comment by npomfret** [comment]
`ExchangeFactory.INSTANCE.createExchange(specification)`

It's a bit confusing known which bit of the text provided by the exchange is the api key and which bit is the secret.

**Comment by npomfret** [comment]
the value of `QuoineExchange.KEY_TOKEN_ID` is the id of the key, and the text lump the give is the secret.

**Comment by paul-c-r** [comment]
Hi guys, I have tried everything suggested above. No luck!
Has anyone found a solution for this?

**Comment by npomfret** [comment]
The info above is correct and works. Use

```
specification.setExchangeSpecificParametersItem(QuoineExchange.KEY_TOKEN_ID, theId);
specification.setExchangeSpecificParametersItem(QuoineExchange.KEY_USER_SECRET, theSecret);
```

**Comment by paul-c-r** [comment]
Thanks Nick, I finally got it working with the help of this trail.
My big mistake was that I instantiated the exchange with the name of the class and then set the specifications against it. As mentioned above, the proper way is to create it with the specs.
I also had some class not found issues for common base64 and utils. I had to add common codec and jack my common-io version up to 2.5
Now it runs like a dream.
Thank you all. I am a very happy chappy!
